File: vb_net/mugwump/mugwump/mugwump.vb
Option Strict On

Imports System

Module mugwump

    '  MUGWUMP
    ' =======
    ' CREATIVE COMPUTING  MORRISTOWN, New JERSEY
    ' Courtesy People's Computer Company
    ' 
    ' Adapted by SeggiePants from the book "BASIC COMPUTER GAMES" Edited by David H. Ahl
    ' 
    ' Can be found online at:  https://www.atariarchives.org/basicgames/showpage.php?page=114

    ' Global Constants
    Const NUM_MUGWUMPS As Integer = 4
    Const GRID_W As Integer = 10
    Const GRID_H As Integer = GRID_W

    Dim Pos(NUM_MUGWUMPS - 1) As position 'Looks like Dim goes from 0 to NUM not 0 to NUM - 1 so subtract one.

    Sub Main(args As String())
        Dim delim() As Char = {","c}
        Dim distance As Double
        Dim i As Integer
        Dim items() As String
        Dim line As String
        Dim playAgain As String
        Dim remaining As Integer
        Dim turn As Integer
        Dim x As Integer
        Dim y As Integer

        For i = 0 To Pos.Length - 1
            Pos(i) = New position()
        Next

        ' Show the introductory text.
        PrintIntroduction()

        While True 'Loop for each game run
            ' Position the MugWumps
            InitMugwumps()
            turn = 1 ' First turn

            While True  'Single Game loop
                Console.WriteLine("")
                Console.WriteLine("")

                ' Get coordinates from the user.
                Do
                    Console.Write(String.Format("Turn No. {0} what is your guess? ", turn))
                    ' Input locations
                    line = Console.ReadLine()
                    items = line.Split(delim)
                    If items.Length >= 2 Then
                        If Not Integer.TryParse(items(0), x) Then
                            x = -1
                        End If

                        If Not Integer.TryParse(items(1), y) Then
                            y = -1
                        End If
                    End If

                Loop Until x >= 0 And x < GRID_W And y >= 0 And y < GRID_H

                ' Check to see if you found a mugwump, report how far away you are otherwise.
                For i = 0 To Pos.Length - 1
                    If Not Pos(i).found Then
                        If Pos(i).x <> x Or Pos(i).y <> y Then
                            distance = Math.Sqrt((Pos(i).x - x) ^ 2 + (Pos(i).y - y) ^ 2)
                            Console.WriteLine("You are {0} units from Mugwump {1}", distance.ToString("0.##"), i + 1, x, y)
                        Else
                            Pos(i).found = True
                            Console.WriteLine(String.Format("YOU HAVE FOUND MUGWUMP {0}", i + 1))
                        End If
                    End If
                Next i

                ' Increment turn And find how many mugwumps remaining.
                turn = turn + 1
                remaining = 0
                For Each mugwump As position  In Pos
                    If Not mugwump.found Then
                        remaining = remaining + 1
                    End If
                Next

                If remaining = 0 Then
                    ' Win State
                    Console.WriteLine()
                    Console.WriteLine(String.Format("YOU GOT THEM ALL IN {0} TURNS!", turn))
                    Exit While
                ElseIf turn >= 10 Then
                    ' Lose State
                    Console.WriteLine()
                    PrettyPrint("SORRY, THAT'S 10 TRIES, HERE is WHERE THEY'RE HIDING")
                    Console.WriteLine()
                    For Each mugwump As position In Pos
                        If mugwump.found = False Then
                            Console.WriteLine(String.Format("MUGWUMP {0} IS AT ({1}, {2})", i + 1, mugwump.x, mugwump.y))
                        End If
                    Next
                    Exit While
                End If
            End While

            ' Would you Like to play again.
            Console.WriteLine()
            Console.WriteLine("THAT WAS FUN!")
            ' Loop until you get Y Or N
            Do
                Console.Write("Would you like to play again? (Y/N) ")
                playAgain = Console.ReadLine().Trim().ToUpper()(0)
            Loop Until playAgain = "Y" Or playAgain = "N"

            ' Break out if you don't want to play anymore.
            If playAgain = "N" Then
                Exit While
            End If

            ' Prepare to restart the game.
            Console.WriteLine("FOUR MORE MUGWUMPS ARE NOW IN HIDING.")
        End While
    End Sub

    ''' <summary>
    ''' Print MESSAGE$ centered on the display.
    ''' </summary>
    ''' <param name="message">The value to print centered on the display.</param>
    ''' 
    Sub CenterPrint(message As String)
        Console.CursorLeft = Math.Max(0, (Console.WindowWidth - message.Length) \ 2)
        Console.WriteLine(message)
    End Sub

    ''' <summary>
    ''' Initialize the position of the MUGWUMPS
    ''' </summary>
    Sub InitMugwumps()
        Dim r As New Random()

        For Each mugwump As position In Pos
            mugwump.RandomPosition(r, GRID_W, GRID_H)
        Next

    End Sub

    ''' <summary>
    ''' Print out a message in a pretty way by splitting characters on word
    ''' boundaries And attempting to keep whole words on a line.
    ''' </summary>
    ''' <param name="message">The message to print.</param>
    Sub PrettyPrint(message As String)
        Dim words() As String
        Dim delim() As Char = {" "c} 'Will let tab, carriage return and line feed pass through as-is

        words = message.Split(delim)
        For Each word As String In words
            If Console.CursorLeft > 0 And Console.CursorLeft < Console.WindowWidth Then
                Console.Write(" ")
            End If
            If Console.CursorLeft + word.Length >= Console.WindowWidth Then
                Console.WriteLine()
            End If
            Console.Write(word)
        Next
    End Sub


    ''' <summary>
    ''' Print the program introduction
    ''' </summary>
    Sub PrintIntroduction()
        Console.Clear()
        CenterPrint("MUGWUMP")
        CenterPrint("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
        Console.WriteLine()
        Console.WriteLine()
        Console.WriteLine()

        ' Courtesy People's Computer Company
        PrettyPrint("The object of this game is to find four Mugwumps ")
        PrettyPrint("hidden on a 10 by 10 grid. Home base is position 0, 0. ")
        PrettyPrint("Any guess you make must be two numbers with each ")
        PrettyPrint("number between 0 and 9 inclusive. The first number ")
        PrettyPrint("is the distance to the right of home base and the second number ")
        PrettyPrint("is distance above home base.")
        Console.WriteLine()
        Console.WriteLine()
        PrettyPrint("You get 10 tries. After each try, I will tell ")
        PrettyPrint("you how far you are from each Mugwump.")
        Console.WriteLine()
        Console.WriteLine()
    End Sub

End Module

File: java/mugwump/mugwump/mugwump/Mugwump.java
package mugwump;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Random;

/**
 * MUGWUMP
 * Find four mugwumps on a 10x10 grid.
 * @author BASIC COMPUTER GAMES edited by David H. Ahl. Ported to Java by SeggiePants
 *
 */


public class Mugwump {
	
	private static final int CONSOLE_W = 80;
	private static final int GRID_W = 10;
	private static final int GRID_H = GRID_W;
	private static final int MAX_MUGWUMPS = 4;
	private static final int MAX_TURNS = 10;
	
	/**
	 * System entry point to the game.
	 * @param args Arguments to the program (not used).
	 */
	public static void main(String[] args)
	{
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		Position[] mugwumps = new Position[MAX_MUGWUMPS];
		String line;
		int remaining;
		int turn;
		int x, y;
		for(int i = 0; i < mugwumps.length; i++)
		{
			mugwumps[i] = new Position();
		}
		
		print_introduction();
		while(true)	// Iterate per game.
		{
			turn = 0;
			init_mugwumps(mugwumps);
			while(true) // Iterate per turn.
			{
				turn++;
				System.out.println();
				System.out.println();
				do {
					System.out.print(String.format("Turn No. %d, what is your guess? ",  turn));
					try {
						line = in.readLine();
					} 
					catch (IOException e)
					{
						line = "";
					}
					String[] numbers = line.split(",");
					if (numbers.length >= 2)
					{
						try
						{
							x = Integer.parseInt(numbers[0].trim());
							y = Integer.parseInt(numbers[1].trim());
						}
						catch (NumberFormatException ex)
						{
							x = -1;
							y = -1;
						}
					}
					else
					{
						x = -1;
						y = -1;
					}
				} while ((x < 0) || (x >= GRID_W) || (y < 0) || (y >= GRID_H));
								
				remaining = 0;
				for (int i = 0; i < mugwumps.length; i++) 
				{
					if (!mugwumps[i].getFound()) 
					{
						if ((mugwumps[i].getX() != x) || (mugwumps[i].getY() != y))
						{
							double distance = Math.sqrt(Math.pow((mugwumps[i].getX() - x), 2) + Math.pow((mugwumps[i].getY() - y), 2));
							System.out.println(String.format("You are %1.2f units from mugwump %d", distance, i + 1));
							remaining++;
						}
						else
						{
							mugwumps[i].setFound(true);
							System.out.println(String.format("YOU HAVE FOUND MUGWUMP %d",  i + 1));							
						}
					}
					
				}
				
				if (remaining == 0)
				{
					System.out.println(String.format("YOU GOT THEM ALL IN %d TURNS.",  turn));
					break;
				}
				else if (turn >= MAX_TURNS) 
				{
					System.out.println();
					System.out.println(String.format("Sorry, that's %d tries. Here is where they're hiding.",  turn));
					for(int i = 0; i < mugwumps.length; i++)
					{
						if (!mugwumps[i].getFound())
						{
							System.out.println(String.format("Mugwump %d is at (%d, %d)", i, mugwumps[i].getX(), mugwumps[i].getY()));
						}
					}
					break;
				}
				
			}
			System.out.println("THAT WAS FUN!");
			do
			{
				System.out.print("Would you like to play again (Y/N)? ");
				try {
					line = in.readLine();
				} 
				catch (IOException e)
				{
					line = "";
				}
				line = line.trim().toUpperCase();				
			} while ((line.length() == 0) || ((!line.substring(0, 1).equals("Y")) && (!line.substring(0, 1).equals("N"))));
			
			if (line.substring(0, 1).equals("N"))
			{
				break;
			}

		}		
		
	}
	
	/**
	 * Print out a message to the console centered on the screen.
	 * @param message The message to print centered. Currently assumes an 80 column console width
	 */
	private static void center_print(String message) 
	{
		int left = Math.max(0,  (CONSOLE_W - message.length()) / 2);
		for(int i = 0; i < left; i++) {
			System.out.print(' ');
		}
		System.out.println(message);		
	}
	
	/**
	 * Position the mugwumps randomly on the grid (they may overlap).
	 * @param mugwumps array of position objects to place on the grid.
	 */
	private static void init_mugwumps(Position[] mugwumps)
	{
		Random r = new Random();
		for(Position mugwump: mugwumps)
		{
			mugwump.RandomPosition(r, GRID_W, GRID_H);
		}
	}
	
	/**
	 * Print out a string to the console taking care to wrap on word boundaries.
	 * @param left The current cursor x-coordinate on the console screen.
	 * @param message The message to print on the screen in a "pretty" way.
	 * @return updated x-coordinate on the screen. should be used like left = pretty_print(left, "value");
	 */
	private static int pretty_print(int left, String message)
	{
		for (String word : message.split("\\s"))
		{
			if (left > 0 && left < CONSOLE_W)
			{
				System.out.print(" ");
				left++;
			}
			
			if (left + word.length() > CONSOLE_W)
			{
				System.out.println();
				left = 0;
			}
			
			System.out.print(word);
			left += word.length();
		}
		return left;
	}
	
	/**
	 * Print out the introduction to the program.
	 */
	private static void print_introduction() 
	{
		center_print("MUGWUMP");
		center_print("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
		System.out.println();
		System.out.println();
		System.out.println();
		// Courtesy People's Computer Company
		int left = 0;
		left = pretty_print(left, "The object of this game is to find four mugwumps.");
		left = pretty_print(left, "hidden on a 10 by 10 grid. Home base is position 0, 0");
		left = pretty_print(left, "any guess you make must be two numbers with each");
		left = pretty_print(left, "number between 0, and 9, inclusive. First number");
		left = pretty_print(left, "is distance to right of home base, and second number");
		left = pretty_print(left, "is distance above home base.");
		System.out.println();
		System.out.println();
		left = 0;
		left = pretty_print(left, "You get 10 tries. After each try, I will tell");
		left = pretty_print(left, "you how far you are from each mugwump.");
		System.out.println();
		System.out.println();
	}
}
